src/main/java/com/google/devtools/build/lib/bazel/bzlmod/VendorUtil.java

```diff
@@ -0,0 +1,162 @@
+package com.google.devtools.build.lib.bazel.bzlmod;
+
+import static java.nio.charset.StandardCharsets.UTF_8;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.hash.HashCode;
+import com.google.common.hash.Hasher;
+import com.google.devtools.build.lib.bazel.repository.downloader.Checksum;
+import com.google.devtools.build.lib.cmdline.RepositoryName;
+import com.google.devtools.build.lib.vfs.FileSystemUtils;
+import com.google.devtools.build.lib.vfs.Path;
+import com.google.devtools.build.lib.vfs.Symlinks;
+
+import java.io.IOException;
+import java.io.UnsupportedEncodingException;
+import java.net.URL;
+import java.net.URLDecoder;
+import java.util.Objects;
+import java.util.Optional;
+
+
+/**
+ * Utility class for vendoring external repositories.
+ */
+public class VendorUtil {
+
+  private final Path vendorDirectory;
+
+  public VendorUtil(Path vendorDirectory) {
+    this.vendorDirectory = vendorDirectory;
+  }
+
+  /**
+   * Vendors the specified repositories under the vendor directory.
+   *
+   * @param externalRepoRoot The root directory of the external repositories.
+   * @param reposToVendor The list of repositories to vendor.
+   * @throws IOException if an I/O error occurs.
+   */
+  public void vendorRepos(
+      Path externalRepoRoot,
+      ImmutableList<RepositoryName> reposToVendor)
+      throws IOException {
+    if (!vendorDirectory.exists()) {
+      vendorDirectory.createDirectoryAndParents();
+    }
+
+    for (RepositoryName repo : reposToVendor) {
```

Comment:
This could probably be parallelized across all repos. 

Do we need to worry about atomicity, e.g. if users interrupt the vendor command? Maybe we should atomically move the marker files into their final location at the end?

Reply:
Indeed, I left a TODO for parallelizing. 

But I think copying one marker file after vendoring one repo source is fine? If the vendoring is interrupted, the user probably doesn't want to re-vendor what's already been vendored?

Reply:
I see! Updated the code to copy the marker file to a tmp file first and then rename it.

Reply:
Yeah moving + leaving a symlink sounds good to me. In fact I think that was my expectation in the beginning, before we settled on copying (can't remember why now).

Reply:
Symlinking does make a small semantic difference: If `external/A` is vendored and thus symlinked to `~/workspace/vendor/A`, but `external/B` is not, then `external/A/../B` will point to `~/workspace/vendor/B`, which doesn't exist. This may turn out fine, but it may also end up breaking some repo rules.

Reply:
> then external/A/../B will point to ~/workspace/vendor/B

This only happens when the symlink is resolved somehow, right? Do people actually use this pattern to locate another repo? I would consider this is a hack?

Reply:
I have used something like this in the past to point `--repo_env=JAVA_HOME` to the embedded JDK for r_j_e's coursier and I could see some repo rules doing this, but I do agree it's a hack :-) I just wanted to point out that we should have such breakages on our radar when we make this optimization (I think we still should).

Reply:
> Yeah moving + leaving a symlink sounds good to me. 

I ended up not doing this. It could cause problem in this situation:
- 1. vendor a repo foo: now external/foo -> vendor_dir/foo
- 2. Remove --vendor_dir
- 3. Think vendor mode is off, and I can change vendor_dir/foo freely
- 4. Build breaks because external/foo still points to the modified source

Reply:
did you actually run into this scenario? I'd expect step 2 to cause RepositoryDelegatorFunction to re-run and realize that --vendor_dir is now off.

Reply:
Yeah, then it will check the marker file, consider it's up-to-date and won't do anything. I guess we can remove the marker file under external/ to make sure the repo is re-refetched when vendor mode is off. I can explore in a next PR.